documentation/bolt_types_reference.md

```diff
@@ -0,0 +1,319 @@
+# Bolt data types
+
+## `ApplyResult`
+
+An `ApplyResult` is returned from an [apply action](https://puppet.com/docs/bolt/latest/applying_manifest_blocks.html#)
+as part of a `ResultSet` object and contains information about the apply action.
+
+The following functions are available to `ApplyResult` objects. They are accessed with
+dot notation `$apply_result.function`.
+
+### `action`
+
+**Returns**
+
+`String`: The type of result (`apply`)
+
+### `error`
+
+**Returns**
+
+`Error`: An object constructed from the `_error` field of the result's value.
+
+### `message`
+
+**Returns**
+
+`String`: The `_output` field of the result's value.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether the result was successful.
+
+### `report`
+
+**Returns**
+
+`Hash`: The Puppet report from the apply action.
+
+### `target`
+
+**Returns**
+
+`Target`: The target the result is from.
+
+### `to_data`
+
+**Returns**
+
+`Hash`: A serialized representation of `ApplyResult`.
+
+
+## `Result`
+
+A `Result` is returned from an action executed on a `Target` as part of a `ResultSet`
+object and contains information about the execution.
+
+The following functions are available to `Result` objects. They are accessed with
+dot notation `$result.function`.
+
+### `[]`
+
+Accesses the value hash directly and returns the value for the key.
+
+> **Note:** This function does not use dot notation and is called directly on the
+  `Result`, i.e. `$result[key]`
+
+**Returns**
+
+`Data`: The accessed value.
+
+### `action`
+
+**Returns**
+
+`String`: The type of result (`task`, `command`, etc.)
+
+### `error`
+
+**Returns**
+
+`Error`: An object constructed from the `_error` field of the result's value.
+
+### `message`
+
+**Returns**
+
+`String`: The `_output` field of the result's value.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether the result was successful.
+
+### `status`
+
+**Returns**
+
+`String`: Either `success` if the result was successful or `failure`.
+
+### `target`
+
+**Returns**
+
+`Target`: The target the result is from.
+
+### `value`
+
+**Returns**
+
+`Hash`: The value of the result.
```

Comment:
This doesn't feel helpful. Maybe "The result of executing the action on the target"